### ISSUE TITLE
Hanging on deleting agent's working folder caused by virusscanner access lock

### ISSUE THREAD

**Issue by gitnubster** [microsoft/azure-pipelines-agent#4089]
We experience issues sometimes when a pipeline is initializing, mostly after the pipeline has ran very recently before. We've found that Windows Defender has set an access lock on a very specific file (a 100mb+ .jar file) that is in the job's working directory of the devops services onprem agent. Apparently no logs were to be found in the agent logs, nor Windows Defender, nor the system's eventlogs. The locked file was only made visible by checking opened handles via sysinternals process explorer.

I would expect the agent to timeout very quickly on this locked file while trying to delete the working directory, but it does not. At least not very fast. It just waits, until the job times out or manual cancellation kicks in. 
Of course we're able to work around this case by excluding the file being locked by Defender, but this situation is not desirable either way. 

That said, this is probably an OS "issue" (and a Defender issue), as manually deleting the same agent working directory makes Windows Explorer unresponsive also. But I'm curious if there is really no way to quick fail on this situation.
 
This is the place where the initialization step seems to hang on in our situation.
https://github.com/microsoft/azure-pipelines-agent/blob/7d3b8d0730ed541fe82523bb81b6b1eaa63653b9/src/Agent.Worker/Build/BuildDirectoryManager.cs#L245

**Comment by sergey-koryshev** [comment]
Hi @gitnubster, if Windows Defender locks file then it considers the file as a malware or some issue occurs during antivirus scan. After which task the file becomes locked? If putting the file into Windows Defender's exclusions is not desirable way but you are sure that the file is trustable one, so you can unlock the file by additional script after affected task I think.

**Comment by gitnubster** [comment]
Hi @sergey-koryshev, we're still trying to get answers and are not sure why the file is an issue. It might be a kind of a deadlock situation on OS level. Only machine reboots help to unlock the file. I'll share new info when I have it.

**Comment by sergey-koryshev** [comment]
@gitnubster, could you please share details about what's going on with Windows Defender when you experience the issue? Does it just scan file or consider it as a threat and wait for user to take some action?

**Comment by gitnubster** [comment]
@sergey-koryshev We've moved on actually. It was not a file that would get quarantined if we scanned it manually, so it was just sitting there under msdefender waiting for something else. We have removed the file from the checked out repo, now the issue is gone. Purely because of that specific file. Only can specify it was a jar file. But that is more like a workaround, and not actually a solved issue. I think we cannot solve it within the ado agent code, as I said it is something at another level/process. Shall I close this issue then?
